### PR TITLE
Allow editors to edit lectures and units

### PR DESCRIPTION
**Pull request by joschahenningsen**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).

### Motivation and Context

**Editors** should be able to **edit** lectures and their corresponding units.

### Steps for Testing

1. Log in as Editor, check if you can to the following:
2. Update, create, delete lectures
3. Update, create, delete, reorder lecture units and attatchments
4. Verify that Instructors/Admins can still do the same

### CODE REVIEW

**Review by ivohashamov — CHANGES_REQUESTED** [review]
Tested on TS2. I tried editing a course in which the user was an editor and the lectures button on the course overview/detail page was visible. However, when on I'm the create lecture page, I'm unable to save new lectures - a Forbidden error code gets returned when clicking on the Save button.

I tested in FloriaGlombikTestkurs with artemis_test_user_10

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 1]
```diff
@@ -71,14 +77,14 @@ public LectureResource(LectureRepository lectureRepository, LectureService lectu
...
-    @PreAuthorize("hasRole('INSTRUCTOR')")
+    @PreAuthorize("hasRole('EDITOR')")
     public ResponseEntity<Lecture> createLecture(@RequestBody Lecture lecture) throws URISyntaxException {
         log.debug("REST request to save Lecture : {}", lecture);
         if (lecture.getId() != null) {
             throw new BadRequestAlertException("A new lecture cannot already have an ID", ENTITY_NAME, "idexists");
         }
         User user = userRepository.getUserWithGroupsAndAuthorities();
-        if (!authCheckService.isInstructorInCourse(lecture.getCourse(), user) && !authCheckService.isAdmin(user)) {
+        if (!(authCheckService.isEditorInCourse(lecture.getCourse(), user) || authCheckService.isInstructorInCourse(lecture.getCourse(), user) || authCheckService.isAdmin(user))) {
```
Comment (FuchsDominik) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:87]:
```suggestion
        if (!authCheckService.isAtLeastEditorInCourse(lecture.getCourse(), user)) {
```
I think we could get rid of these separate authorization checks and simply use the frequently used `isAtLeastXXX` check, which shortens the if clause and handles superior roles as well :)

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 2]
```diff
@@ -95,14 +101,14 @@ public LectureResource(LectureRepository lectureRepository, LectureService lectu
...
-    @PreAuthorize("hasRole('INSTRUCTOR')")
+    @PreAuthorize("hasRole('EDITOR')")
     public ResponseEntity<Lecture> updateLecture(@RequestBody Lecture lecture) throws URISyntaxException {
         log.debug("REST request to update Lecture : {}", lecture);
         if (lecture.getId() == null) {
             throw new BadRequestAlertException("Invalid id", ENTITY_NAME, "idnull");
         }
         User user = userRepository.getUserWithGroupsAndAuthorities();
-        if (!authCheckService.isInstructorInCourse(lecture.getCourse(), user) && !authCheckService.isAdmin(user)) {
+        if (!(authCheckService.isEditorInCourse(lecture.getCourse(), user) || authCheckService.isInstructorInCourse(lecture.getCourse(), user) || authCheckService.isAdmin(user))) {
```
Comment (FuchsDominik) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:111]:
```suggestion
        if (!authCheckService.isAtLeastEditorInCourse(lecture.getCourse(), user)) {
```
Here as well

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 3]
```diff
@@ -124,13 +130,13 @@ public LectureResource(LectureRepository lectureRepository, LectureService lectu
...
     @GetMapping(value = "/courses/{courseId}/lectures")
-    @PreAuthorize("hasRole('INSTRUCTOR')")
+    @PreAuthorize("hasRole('EDITOR')")
     public ResponseEntity<Set<Lecture>> getLecturesForCourse(@PathVariable Long courseId, @RequestParam(required = false, defaultValue = "false") boolean withLectureUnits) {
         log.debug("REST request to get all Lectures for the course with id : {}", courseId);
 
         User user = userRepository.getUserWithGroupsAndAuthorities();
         Course course = courseRepository.findByIdElseThrow(courseId);
-        if (!authCheckService.isInstructorInCourse(course, user) && !authCheckService.isAdmin(user)) {
+        if (!(authCheckService.isEditorInCourse(course, user) || authCheckService.isInstructorInCourse(course, user) || authCheckService.isAdmin(user))) {
```
Comment (FuchsDominik) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:139]:
```suggestion
        if (!authCheckService.isAtLeastEditorInCourse(lecture.getCourse(), user)) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 4]
```diff
@@ -231,15 +237,15 @@ else if (lectureUnit instanceof AttachmentUnit) {
...
+    @PreAuthorize("hasRole('EDITOR')")
     public ResponseEntity<Void> deleteLecture(@PathVariable Long id) {
         User user = userRepository.getUserWithGroupsAndAuthorities();
         Optional<Lecture> optionalLecture = lectureRepository.findByIdWithStudentQuestionsAndLectureUnitsAndLearningGoals(id);
         if (optionalLecture.isEmpty()) {
             return ResponseEntity.notFound().build();
         }
         Lecture lecture = optionalLecture.get();
-        if (!authCheckService.isInstructorInCourse(lecture.getCourse(), user) && !authCheckService.isAdmin(user)) {
+        if (!(authCheckService.isEditorInCourse(lecture.getCourse(), user) || authCheckService.isInstructorInCourse(lecture.getCourse(), user) || authCheckService.isAdmin(user))) {
```
Comment (FuchsDominik) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:248]:
```suggestion
        if (!authCheckService.isAtLeastEditorInCourse(lecture.getCourse(), user)) {
```

**Review by FuchsDominik — CHANGES_REQUESTED** [review]
Code looks good :) Only some small remarks

**Review by katjanakosic — APPROVED** [review]
Tested on TS1 and everything works as expected. I had no issues.

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java (hunk 1 above)
Comment (joschahenningsen) [reply src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java]: Good idea, thanks!

File: src/main/java/de/tum/in/www1/artemis/service/AuthorizationCheckService.java [hunk 5]
```diff
@@ -533,7 +533,7 @@ public boolean isAllowedToSeeLectureUnit(@NotNull LectureUnit lectureUnit, @Null
             return true;
         }
         Course course = lectureUnit.getLecture().getCourse();
-        return isInstructorInCourse(course, user) || isTeachingAssistantInCourse(course, user) || (isStudentInCourse(course, user) && lectureUnit.isVisibleToStudents());
+        return isEditorInCourse(course, user) || isInstructorInCourse(course, user) || isTeachingAssistantInCourse(course, user) || (isStudentInCourse(course, user) && lectureUnit.isVisibleToStudents());
```
Comment (FuchsDominik) [inline src/main/java/de/tum/in/www1/artemis/service/AuthorizationCheckService.java:536]:
```suggestion
        return isAtLeastTeachingAssistantInCourse(course, user) || (isStudentInCourse(course, user) && lectureUnit.isVisibleToStudents());
```

Comment (FuchsDominik) [reply src/main/java/de/tum/in/www1/artemis/service/AuthorizationCheckService.java:536]: My bad, i missed this spot

**Review by canyelten — CHANGES_REQUESTED** [review]
Tested on TS1. Instructors can still update, create, delete, reorder lecture units and attachments.  But the Lectures button wasn't visible in the Course management view so I couldn't do anything about lectures as an editor. In the course overview view,  students were able to see the lecture.

**Comment by joschahenningsen** [comment]
> But the Lectures button wasn't visible in the Course management view so I couldn't do anything about lectures as an editor.

@canyelten can't reproduce this right now, which account and test course did you use for testing?

**Review by maximilianbergmann — APPROVED** [review]
Tested on TS3. As an editor I could create a lecture, add and delete elements in the lecture. Instructor perspective works as always. Student perspective ok.

**Comment by canyelten** [comment]
> > But the Lectures button wasn't visible in the Course management view so I couldn't do anything about lectures as an editor.
> 
> @canyelten can't reproduce this right now, which account and test course did you use for testing?

**Comment by canyelten** [comment]
> > But the Lectures button wasn't visible in the Course management view so I couldn't do anything about lectures as an editor.
> 
> @canyelten can't reproduce this right now, which account and test course did you use for testing?

I used TU11 as an instructor and gave editor rights to TU2 for my own test course "Can Yelten Test Course"

**Comment by joschahenningsen** [comment]
> I used TU11 as an instructor and gave editor rights to TU2 for my own test course "Can Yelten Test Course"

I can see the lectures button using tu2.

**Comment by canyelten** [comment]
> > I used TU11 as an instructor and gave editor rights to TU2 for my own test course "Can Yelten Test Course"
> 
> I can see the lectures button using tu2.

That's weird, probably I did something wrong:(

**Comment by joschahenningsen** [comment]
@canyelten does logging out and back in with the user help?

**Review by willergeorg — APPROVED** [review]
Tested on TS1. No errors occurred.

**Review by NicolasUllmann — APPROVED** [review]
Tested on Server 2. Everything worked as expected.

**Comment by canyelten** [comment]
> @canyelten does logging out and back in with the user help?
Yes, it helped. I can now update, create, delete lectures as an editor: the button on the course management page is visible. But I have noticed that the button is still missing when I click on the course.

**Comment by joschahenningsen** [comment]
@canyelten Thanks for noticing! Pushed the fix.

**Review by FuchsDominik — APPROVED** [review]
Change requests got addressed 👍  Code-wise approve

**Review by canyelten — CHANGES_REQUESTED** [review]
Tested on TS1 using TU2 and TU3. The lectures button is still missing when I click on the course as Editor.

**Review by krusche — APPROVED** [review]

**Pull request merged**